flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/ResultPartitionFactory.java

```diff
@@ -253,6 +348,46 @@ public ResultPartition create(
         return partition;
     }
 
+    private List<TieredStorageMemorySpec> createTieredStorageMemorySpecs(
+            int numberOfSubpartitions,
+            BufferAccumulator bufferAccumulator,
+            List<TierProducerAgent> tierProducerAgents) {
+        List<TieredStorageMemorySpec> tieredStorageMemorySpecs = new ArrayList<>();
+        tieredStorageMemorySpecs.add(
+                new TieredStorageMemorySpec(
+                        bufferAccumulator,
+                        Math.min(
+                                numberOfSubpartitions + 1,
+                                checkNotNull(tieredStorageConfiguration)
+                                        .getAccumulatorExclusiveBuffers())));
+        for (TierProducerAgent tierProducerAgent : tierProducerAgents) {
+            if (tierProducerAgent.getClass() == MemoryTierProducerAgent.class) {
+                tieredStorageMemorySpecs.add(
+                        new TieredStorageMemorySpec(
+                                tierProducerAgent,
+                                checkNotNull(tieredStorageConfiguration)
+                                        .getMemoryTierExclusiveBuffers()));
+            }
+
+            if (tierProducerAgent.getClass() == DiskTierProducerAgent.class) {
+                tieredStorageMemorySpecs.add(
+                        new TieredStorageMemorySpec(
+                                tierProducerAgent,
+                                checkNotNull(tieredStorageConfiguration)
+                                        .getDiskTierExclusiveBuffers()));
+            }
+
+            if (tierProducerAgent.getClass() == RemoteTierProducerAgent.class) {
+                tieredStorageMemorySpecs.add(
+                        new TieredStorageMemorySpec(
+                                tierProducerAgent,
+                                checkNotNull(tieredStorageConfiguration)
+                                        .getRemoteTierExclusiveBuffers()));
+            }
```

Comment:
This doesn't seem right. We should not bind the memory spec configuration to the agent classes. Instead, we should create both the factories and memory specs from configuration, possibility in `TieredStorageConfiguration#createDefaultFactories`.